### PR TITLE
Add unit tests for libexpr

### PR DESCRIPTION
**Pull request by andir**
This is an initial step towards working on the libexpr code and trying
to optimize it. Having unit tests will be a great starting point to
write a few benchmarks that focus on various features of NIx. Those
might not have to be part of the code base but are likely very helpful
while developing.

### CODE REVIEW

File: src/libexpr/tests/primops.cc [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+namespace nix {
+    class LibExprTest : public ::testing::Test {
+        protected:
+            static void SetUpTestSuite() {
+                initGC();
+            }
+
+            LibExprTest()
+                : store(openStore("dummy://"))
+                , state({"."}, store)
```
[the remark this answers is not in the document]
Comment (andir) [reply src/libexpr/tests/primops.cc:22]: Could you point out where it is stateful beyond its own lifetime? Happy to start reducing the statefulness of it as part of this work.

File: src/libexpr/tests/primops.cc [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        auto v = eval("-1");
+        ASSERT_EQ(v.type(), nInt);
+        ASSERT_EQ(v.integer, -1);
+    }
+
+    TEST_F(TrivialExpression, 1minus1) {
+        auto v = eval("1-1");
+        ASSERT_EQ(v.type(), nInt);
+        ASSERT_EQ(v.integer, 0);
+    }
```
[the remark this answers is not in the document]
Comment (andir) [reply src/libexpr/tests/primops.cc:88]: Yeah, I can see the appeal in those. Until I can figure out why it is crashing I'd rather not dig into that rabbit hole just yet.

**Comment by andir** [comment]
Picking up my debugging quest I've found out that the primOps aren't always registered with the correct arity when more than one EvalState is being initialized:

```
[ RUN      ] TrivialExpression.true
Registering primOp __getFlake with arity 1
Registering primOp scopedImport with arity 2
Registering primOp import with arity 1
Registering primOp __typeOf with arity 1
Registering primOp isNull with arity 1
Registering primOp __isFunction with arity 1
Registering primOp __isInt with arity 1
Registering primOp __isFloat with arity 1
Registering primOp __isString with arity 1
Registering primOp __isBool with arity 1
Registering primOp __isPath with arity 1
Registering primOp __genericClosure with arity 1
Registering primOp abort with arity 1
Registering primOp throw with arity 1
Registering primOp __addErrorContext with arity 2
Registering primOp __ceil with arity 1
Registering primOp __floor with arity 1
Registering primOp __tryEval with arity 1
Registering primOp __getEnv with arity 1
Registering primOp __seq with arity 2
Registering primOp __deepSeq with arity 2
Registering primOp __trace with arity 2
Registering primOp derivationStrict with arity 1
Registering primOp placeholder with arity 1
Registering primOp __toPath with arity 1
Registering primOp __storePath with arity 1
Registering primOp __pathExists with arity 1
Registering primOp baseNameOf with arity 1
Registering primOp dirOf with arity 1
Registering primOp __readFile with arity 1
Registering primOp __findFile with arity 2
Registering primOp __hashFile with arity 2
Registering primOp __readDir with arity 1
Registering primOp __toXML with arity 1
Registering primOp __toJSON with arity 1
Registering primOp __fromJSON with arity 1
Registering primOp __toFile with arity 2
Registering primOp __filterSource with arity 2
Registering primOp __path with arity 1
Registering primOp __attrNames with arity 1
Registering primOp __attrValues with arity 1
Registering primOp __getAttr with arity 2
Registering primOp __unsafeGetAttrPos with arity 2
Registering primOp __hasAttr with arity 2
Registering primOp __isAttrs with arity 1
Registering primOp removeAttrs with arity 2
Registering primOp __listToAttrs with arity 1
Registering primOp __intersectAttrs with arity 2
Registering primOp __catAttrs with arity 2
Registering primOp __functionArgs with arity 1
Registering primOp __mapAttrs with arity 2
Registering primOp __isList with arity 1
Registering primOp __elemAt with arity 2
Registering primOp __head with arity 1
Registering primOp __tail with arity 1
Registering primOp map with arity 2
Registering primOp __filter with arity 2
Registering primOp __elem with arity 2
Registering primOp __concatLists with arity 1
Registering primOp __length with arity 1
Registering primOp __foldl' with arity 3
Registering primOp __any with arity 2
Registering primOp __all with arity 2
Registering primOp __genList with arity 2
Registering primOp __sort with arity 2
Registering primOp __partition with arity 2
Registering primOp __concatMap with arity 2
Registering primOp __add with arity 2
Registering primOp __sub with arity 2
Registering primOp __mul with arity 2
Registering primOp __div with arity 2
Registering primOp __bitAnd with arity 2
Registering primOp __bitOr with arity 2
Registering primOp __bitXor with arity 2
Registering primOp __lessThan with arity 2
Registering primOp toString with arity 1
Registering primOp __substring with arity 3
Registering primOp __stringLength with arity 1
Registering primOp __hashString with arity 2
Registering primOp __match with arity 2
Registering primOp __split with arity 2
Registering primOp __concatStringsSep with arity 2
Registering primOp __replaceStrings with arity 3
Registering primOp __parseDrvName with arity 1
Registering primOp __compareVersions with arity 2
Registering primOp __splitVersion with arity 1
Registering primOp __unsafeDiscardStringContext with arity 1
Registering primOp __hasContext with arity 1
Registering primOp __unsafeDiscardOutputDependency with arity 1
Registering primOp __getContext with arity 1
Registering primOp __appendContext with arity 2
Registering primOp fetchMercurial with arity 1
Registering primOp fetchTree with arity 1
Registering primOp __fetchurl with arity 1
Registering primOp fetchTarball with arity 1
Registering primOp fetchGit with arity 1
Registering primOp fromTOML with arity 1
[       OK ] TrivialExpression.true (12 ms)
[ RUN      ] TrivialExpression.false
Registering primOp __getFlake with arity 1
Registering primOp scopedImport with arity 2
Registering primOp import with arity 0
Registering primOp __typeOf with arity 0
Registering primOp isNull with arity 0
Registering primOp __isFunction with arity 0
Registering primOp __isInt with arity 0
Registering primOp __isFloat with arity 0
Registering primOp __isString with arity 0
Registering primOp __isBool with arity 0
Registering primOp __isPath with arity 0
Registering primOp __genericClosure with arity 1
Registering primOp abort with arity 0
Registering primOp throw with arity 0
Registering primOp __addErrorContext with arity 2
Registering primOp __ceil with arity 0
Registering primOp __floor with arity 0
Registering primOp __tryEval with arity 0
Registering primOp __getEnv with arity 0
Registering primOp __seq with arity 0
Registering primOp __deepSeq with arity 0
Registering primOp __trace with arity 0
Registering primOp derivationStrict with arity 1
Registering primOp placeholder with arity 0
Registering primOp __toPath with arity 0
Registering primOp __storePath with arity 0
Registering primOp __pathExists with arity 0
Registering primOp baseNameOf with arity 0
Registering primOp dirOf with arity 0
Registering primOp __readFile with arity 0
Registering primOp __findFile with arity 2
Registering primOp __hashFile with arity 0
Registering primOp __readDir with arity 0
Registering primOp __toXML with arity 0
Registering primOp __toJSON with arity 0
Registering primOp __fromJSON with arity 0
Registering primOp __toFile with arity 0
Registering primOp __filterSource with arity 0
Registering primOp __path with arity 0
Registering primOp __attrNames with arity 0
Registering primOp __attrValues with arity 0
Registering primOp __getAttr with arity 0
Registering primOp __unsafeGetAttrPos with arity 2
Registering primOp __hasAttr with arity 0
Registering primOp __isAttrs with arity 0
Registering primOp removeAttrs with arity 0
Registering primOp __listToAttrs with arity 0
Registering primOp __intersectAttrs with arity 0
Registering primOp __catAttrs with arity 0
Registering primOp __functionArgs with arity 0
Registering primOp __mapAttrs with arity 0
Registering primOp __isList with arity 0
Registering primOp __elemAt with arity 0
Registering primOp __head with arity 0
Registering primOp __tail with arity 0
Registering primOp map with arity 0
Registering primOp __filter with arity 0
Registering primOp __elem with arity 0
Registering primOp __concatLists with arity 0
Registering primOp __length with arity 0
Registering primOp __foldl' with arity 0
Registering primOp __any with arity 0
Registering primOp __all with arity 0
Registering primOp __genList with arity 0
Registering primOp __sort with arity 0
Registering primOp __partition with arity 0
Registering primOp __concatMap with arity 0
Registering primOp __add with arity 0
Registering primOp __sub with arity 0
Registering primOp __mul with arity 0
Registering primOp __div with arity 0
Registering primOp __bitAnd with arity 0
Registering primOp __bitOr with arity 0
Registering primOp __bitXor with arity 0
Registering primOp __lessThan with arity 0
Registering primOp toString with arity 0
Registering primOp __substring with arity 0
Registering primOp __stringLength with arity 0
Registering primOp __hashString with arity 0
Registering primOp __match with arity 0
Registering primOp __split with arity 0
Registering primOp __concatStringsSep with arity 0
Registering primOp __replaceStrings with arity 0
Registering primOp __parseDrvName with arity 0
Registering primOp __compareVersions with arity 0
Registering primOp __splitVersion with arity 0
Registering primOp __unsafeDiscardStringContext with arity 1
Registering primOp __hasContext with arity 1
Registering primOp __unsafeDiscardOutputDependency with arity 1
Registering primOp __getContext with arity 1
Registering primOp __appendContext with arity 2
Registering primOp fetchMercurial with arity 1
Registering primOp fetchTree with arity 1
Registering primOp __fetchurl with arity 0
Registering primOp fetchTarball with arity 0
Registering primOp fetchGit with arity 0
Registering primOp fromTOML with arity 1
[       OK ] TrivialExpression.false (1 ms)
[ RUN      ] TrivialExpression.null
Registering primOp __getFlake with arity 1
Registering primOp scopedImport with arity 2
Registering primOp import with arity 0
Registering primOp __typeOf with arity 0
Registering primOp isNull with arity 0
Registering primOp __isFunction with arity 0
Registering primOp __isInt with arity 0
Registering primOp __isFloat with arity 0
Registering primOp __isString with arity 0
Registering primOp __isBool with arity 0
Registering primOp __isPath with arity 0
Registering primOp __genericClosure with arity 1
Registering primOp abort with arity 0
Registering primOp throw with arity 0
Registering primOp __addErrorContext with arity 2
Registering primOp __ceil with arity 0
Registering primOp __floor with arity 0
Registering primOp __tryEval with arity 0
Registering primOp __getEnv with arity 0
Registering primOp __seq with arity 0
Registering primOp __deepSeq with arity 0
Registering primOp __trace with arity 0
Registering primOp derivationStrict with arity 1
Registering primOp placeholder with arity 0
Registering primOp __toPath with arity 0
Registering primOp __storePath with arity 0
Registering primOp __pathExists with arity 0
Registering primOp baseNameOf with arity 0
Registering primOp dirOf with arity 0
Registering primOp __readFile with arity 0
Registering primOp __findFile with arity 2
Registering primOp __hashFile with arity 0
Registering primOp __readDir with arity 0
Registering primOp __toXML with arity 0
Registering primOp __toJSON with arity 0
Registering primOp __fromJSON with arity 0
Registering primOp __toFile with arity 0
Registering primOp __filterSource with arity 0
Registering primOp __path with arity 0
Registering primOp __attrNames with arity 0
Registering primOp __attrValues with arity 0
Registering primOp __getAttr with arity 0
Registering primOp __unsafeGetAttrPos with arity 2
Registering primOp __hasAttr with arity 0
Registering primOp __isAttrs with arity 0
Registering primOp removeAttrs with arity 0
Registering primOp __listToAttrs with arity 0
Registering primOp __intersectAttrs with arity 0
Registering primOp __catAttrs with arity 0
Registering primOp __functionArgs with arity 0
Registering primOp __mapAttrs with arity 0
Registering primOp __isList with arity 0
Registering primOp __elemAt with arity 0
Registering primOp __head with arity 0
Registering primOp __tail with arity 0
Registering primOp map with arity 0
Registering primOp __filter with arity 0
Registering primOp __elem with arity 0
Registering primOp __concatLists with arity 0
Registering primOp __length with arity 0
Registering primOp __foldl' with arity 0
Registering primOp __any with arity 0
Registering primOp __all with arity 0
Registering primOp __genList with arity 0
Registering primOp __sort with arity 0
Registering primOp __partition with arity 0
Registering primOp __concatMap with arity 0
Registering primOp __add with arity 0
Registering primOp __sub with arity 0
Registering primOp __mul with arity 0
Registering primOp __div with arity 0
Registering primOp __bitAnd with arity 0
Registering primOp __bitOr with arity 0
Registering primOp __bitXor with arity 0
Registering primOp __lessThan with arity 0
Registering primOp toString with arity 0
Registering primOp __substring with arity 0
Registering primOp __stringLength with arity 0
Registering primOp __hashString with arity 0
Registering primOp __match with arity 0
Registering primOp __split with arity 0
Registering primOp __concatStringsSep with arity 0
Registering primOp __replaceStrings with arity 0
Registering primOp __parseDrvName with arity 0
Registering primOp __compareVersions with arity 0
Registering primOp __splitVersion with arity 0
Registering primOp __unsafeDiscardStringContext with arity 1
Registering primOp __hasContext with arity 1
Registering primOp __unsafeDiscardOutputDependency with arity 1
Registering primOp __getContext with arity 1
Registering primOp __appendContext with arity 2
Registering primOp fetchMercurial with arity 1
Registering primOp fetchTree with arity 1
Registering primOp __fetchurl with arity 0
Registering primOp fetchTarball with arity 0
Registering primOp fetchGit with arity 0
Registering primOp fromTOML with arity 1
[       OK ] TrivialExpression.null (1 ms)
[ RUN      ] TrivialExpression.1
Registering primOp __getFlake with arity 1
Registering primOp scopedImport with arity 2
Registering primOp import with arity 0
Registering primOp __typeOf with arity 0
Registering primOp isNull with arity 0
Registering primOp __isFunction with arity 0
Registering primOp __isInt with arity 0
Registering primOp __isFloat with arity 0
Registering primOp __isString with arity 0
Registering primOp __isBool with arity 0
Registering primOp __isPath with arity 0
Registering primOp __genericClosure with arity 1
Registering primOp abort with arity 0
Registering primOp throw with arity 0
Registering primOp __addErrorContext with arity 2
Registering primOp __ceil with arity 0
Registering primOp __floor with arity 0
Registering primOp __tryEval with arity 0
Registering primOp __getEnv with arity 0
Registering primOp __seq with arity 0
Registering primOp __deepSeq with arity 0
Registering primOp __trace with arity 0
Registering primOp derivationStrict with arity 1
Registering primOp placeholder with arity 0
Registering primOp __toPath with arity 0
Registering primOp __storePath with arity 0
Registering primOp __pathExists with arity 0
Registering primOp baseNameOf with arity 0
Registering primOp dirOf with arity 0
Registering primOp __readFile with arity 0
Registering primOp __findFile with arity 2
Registering primOp __hashFile with arity 0
Registering primOp __readDir with arity 0
Registering primOp __toXML with arity 0
Registering primOp __toJSON with arity 0
Registering primOp __fromJSON with arity 0
Registering primOp __toFile with arity 0
Registering primOp __filterSource with arity 0
Registering primOp __path with arity 0
Registering primOp __attrNames with arity 0
Registering primOp __attrValues with arity 0
Registering primOp __getAttr with arity 0
Registering primOp __unsafeGetAttrPos with arity 2
Registering primOp __hasAttr with arity 0
Registering primOp __isAttrs with arity 0
Registering primOp removeAttrs with arity 0
Registering primOp __listToAttrs with arity 0
Registering primOp __intersectAttrs with arity 0
Registering primOp __catAttrs with arity 0
Registering primOp __functionArgs with arity 0
Registering primOp __mapAttrs with arity 0
Registering primOp __isList with arity 0
Registering primOp __elemAt with arity 0
Registering primOp __head with arity 0
Registering primOp __tail with arity 0
Registering primOp map with arity 0
Registering primOp __filter with arity 0
Registering primOp __elem with arity 0
Registering primOp __concatLists with arity 0
Registering primOp __length with arity 0
Registering primOp __foldl' with arity 0
Registering primOp __any with arity 0
Registering primOp __all with arity 0
Registering primOp __genList with arity 0
Registering primOp __sort with arity 0
Registering primOp __partition with arity 0
Registering primOp __concatMap with arity 0
Registering primOp __add with arity 0
Registering primOp __sub with arity 0
Registering primOp __mul with arity 0
Registering primOp __div with arity 0
Registering primOp __bitAnd with arity 0
Registering primOp __bitOr with arity 0
Registering primOp __bitXor with arity 0
Registering primOp __lessThan with arity 0
Registering primOp toString with arity 0
Registering primOp __substring with arity 0
Registering primOp __stringLength with arity 0
Registering primOp __hashString with arity 0
Registering primOp __match with arity 0
Registering primOp __split with arity 0
Registering primOp __concatStringsSep with arity 0
Registering primOp __replaceStrings with arity 0
Registering primOp __parseDrvName with arity 0
Registering primOp __compareVersions with arity 0
Registering primOp __splitVersion with arity 0
Registering primOp __unsafeDiscardStringContext with arity 1
Registering primOp __hasContext with arity 1
Registering primOp __unsafeDiscardOutputDependency with arity 1
Registering primOp __getContext with arity 1
Registering primOp __appendContext with arity 2
Registering primOp fetchMercurial with arity 1
Registering primOp fetchTree with arity 1
Registering primOp __fetchurl with arity 0
Registering primOp fetchTarball with arity 0
Registering primOp fetchGit with arity 0
Registering primOp fromTOML with arity 1
[       OK ] TrivialExpression.1 (0 ms)
[ RUN      ] TrivialExpression.minus1
Registering primOp __getFlake with arity 1
Registering primOp scopedImport with arity 2
Registering primOp import with arity 0
Registering primOp __typeOf with arity 0
Registering primOp isNull with arity 0
Registering primOp __isFunction with arity 0
Registering primOp __isInt with arity 0
Registering primOp __isFloat with arity 0
Registering primOp __isString with arity 0
Registering primOp __isBool with arity 0
Registering primOp __isPath with arity 0
Registering primOp __genericClosure with arity 1
Registering primOp abort with arity 0
Registering primOp throw with arity 0
Registering primOp __addErrorContext with arity 2
Registering primOp __ceil with arity 0
Registering primOp __floor with arity 0
Registering primOp __tryEval with arity 0
Registering primOp __getEnv with arity 0
Registering primOp __seq with arity 0
Registering primOp __deepSeq with arity 0
Registering primOp __trace with arity 0
Registering primOp derivationStrict with arity 1
Registering primOp placeholder with arity 0
Registering primOp __toPath with arity 0
Registering primOp __storePath with arity 0
Registering primOp __pathExists with arity 0
Registering primOp baseNameOf with arity 0
Registering primOp dirOf with arity 0
Registering primOp __readFile with arity 0
Registering primOp __findFile with arity 2
Registering primOp __hashFile with arity 0
Registering primOp __readDir with arity 0
Registering primOp __toXML with arity 0
Registering primOp __toJSON with arity 0
Registering primOp __fromJSON with arity 0
Registering primOp __toFile with arity 0
Registering primOp __filterSource with arity 0
Registering primOp __path with arity 0
Registering primOp __attrNames with arity 0
Registering primOp __attrValues with arity 0
Registering primOp __getAttr with arity 0
Registering primOp __unsafeGetAttrPos with arity 2
Registering primOp __hasAttr with arity 0
Registering primOp __isAttrs with arity 0
Registering primOp removeAttrs with arity 0
Registering primOp __listToAttrs with arity 0
Registering primOp __intersectAttrs with arity 0
Registering primOp __catAttrs with arity 0
Registering primOp __functionArgs with arity 0
Registering primOp __mapAttrs with arity 0
Registering primOp __isList with arity 0
Registering primOp __elemAt with arity 0
Registering primOp __head with arity 0
Registering primOp __tail with arity 0
Registering primOp map with arity 0
Registering primOp __filter with arity 0
Registering primOp __elem with arity 0
Registering primOp __concatLists with arity 0
Registering primOp __length with arity 0
Registering primOp __foldl' with arity 0
Registering primOp __any with arity 0
Registering primOp __all with arity 0
Registering primOp __genList with arity 0
Registering primOp __sort with arity 0
Registering primOp __partition with arity 0
Registering primOp __concatMap with arity 0
Registering primOp __add with arity 0
Registering primOp __sub with arity 0
Registering primOp __mul with arity 0
Registering primOp __div with arity 0
Registering primOp __bitAnd with arity 0
Registering primOp __bitOr with arity 0
Registering primOp __bitXor with arity 0
Registering primOp __lessThan with arity 0
Registering primOp toString with arity 0
Registering primOp __substring with arity 0
Registering primOp __stringLength with arity 0
Registering primOp __hashString with arity 0
Registering primOp __match with arity 0
Registering primOp __split with arity 0
Registering primOp __concatStringsSep with arity 0
Registering primOp __replaceStrings with arity 0
Registering primOp __parseDrvName with arity 0
Registering primOp __compareVersions with arity 0
Registering primOp __splitVersion with arity 0
Registering primOp __unsafeDiscardStringContext with arity 1
Registering primOp __hasContext with arity 1
Registering primOp __unsafeDiscardOutputDependency with arity 1
Registering primOp __getContext with arity 1
Registering primOp __appendContext with arity 2
Registering primOp fetchMercurial with arity 1
Registering primOp fetchTree with arity 1
Registering primOp __fetchurl with arity 0
Registering primOp fetchTarball with arity 0
Registering primOp fetchGit with arity 0
Registering primOp fromTOML with arity 1
```

**Comment by andir** [comment]
As it turned out, the arguments were moved out of the registration structure and thus every initialization after would get the wrong data. If we really care about that small amount of memory overhead, we should put them into a shared_ptr that is cheap to clone. As far as I can tell the evaluator only needs the structure to figure out how many arguments should there be. There is a one time operation. It is probably more of an impact on the new repl `:doc` feature, but a pointer indirection sounds like a reasonable tradeoff there. The simplest fix was to remove the `std::move` invocation. As done in 
9bfc5d2.

**Comment by andir** [comment]
@regnat mind giving this another review?

**Comment by edolstra** [comment]
Note that we already have libexpr tests in `tests/lang`. What's the advantage of testing the evaluator in C++?

**Comment by andir** [comment]
> Note that we already have libexpr tests in `tests/lang`. What's the advantage of testing the evaluator in C++?

I wanted some simple & fast tests that ensure that some basic assumptions around the language continue to hold true. It also serves as a way to ensure that we can eval in multiple separate "environments" (multiple EvalStates) during the runtime of a single process.

The tests currently execut within ~20ms and compile within a couple of seconds. I don't need a full (re)build of Nix executables to verify my changes aren't breaking fundamental assumptions.

I don't think this will replace the tests we have in`tests/lang` but might be an option for more leightweight language feature tests. For example we can ensure that certain constructs are lazy (map) or that other operations. It is also easier (at least for me) to debug segfaults during eval when I can reproduce them in a minimal executable.

The `tests/lang` tests are great for larger expressions that might have failed in the past and that should not fail ever again. These smaller tests are more suited towards testing axioms and/or edge cases.

In the past made several attempts at refactoring various parts in `libexpr` and then ended up with some segfault that would only occur when testing against nixpkgs. Stepping through the eval of nixpkgs isn't really fun in my opinion. The hope is that we could reproduce certain issues with smaller and easier to debug unit tests in those cases.

File: src/libexpr/tests/primops.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (andir) [reply src/libexpr/tests/primops.cc]: I've now used this extensively and with that one primOp fix applied I've not seen any issues. I used this work to fuzz the parser multiple thousand times per second within one process for a few days without issues. If there is still some global state (besides the GC) I can't find it....

File: src/libexpr/tests/json.cc [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        mkInt(v, -100);
+        ASSERT_EQ(getJSONValue(v), "-100");
+    }
+
+    TEST_F(JSONValueTest, String) {
+        Value v;
+        mkString(v, "test");
+        ASSERT_EQ(getJSONValue(v), "\"test\"");
+    }
+
```
Comment (gilligan) [inline src/libexpr/tests/json.cc:52]: Maybe you could also add a test for unicode strings here

**Comment by andir** [comment]
Rebased one more time and added another corner case that is odd in our current language implementation (`or = 1;` is valid but you can't ever access the value bound to it unless you go through something like `x.or`).

**Review by thufschmitt — APPROVED** [review]
@andir Sorry for taking so long to get back to it.

I’ve rebased the branch on top of master. I let you double-check that I didn’t screw anything up in the process if you want, and I think we can merge it afterwards

**Comment by andir** [comment]
> @andir Sorry for taking so long to get back to it.
> 
> I’ve rebased the branch on top of master. I let you double-check that I didn’t screw anything up in the process if you want, and I think we can merge it afterwards

I did a brief casual browsing through the diff and I couldn't see anything that is obviously wrong. Go ahead and merge :-)

**Comment by edolstra** [comment]
I'm not very happy that we now have two different test suites for the evaluator. This just makes it confusing whether tests should go into libexpr-tests or tests/lang. If this PR had replaced tests/lang with libexpr-tests, that would have been one thing, but now it just increases maintainer burden.

**Comment by edolstra** [comment]
The other problem is that libexpr-tests is intimately tied to internal APIs, such as the Value representation. So changes to those internal interfaces (e.g. `Value::listElems()`, to give an example of something that has changed recently) would require a huge amount of work to update the tests - work that would be unnecessary with the test framework in tests/lang.

**Comment by thufschmitt** [comment]
> If this PR had replaced tests/lang with libexpr-tests, that would have been one thing, but now it just increases maintainer burden.

A bit yes. But at the same time I wouldn’t want to commit to it too quickly (esp. wrt the dependency on the internal APIs that you mention: _a priori_, I think it’s a worthwile tradeoff, but I’d rather give it some time too ensure that it doesn’t cause too much pain before making a hard-to-revert change). Keeping it in addition to the existing testsuite makes it so that we have time to see how that’s gonna work out and migrate incrementally what we think needs to be migrated.

> This style of writing tests is also extremely verbose

Yes, it is :) Maybe we could just compare the two values for this kind of more complex examples (either print the result and compare the strings, or use the equality operator from Nix).

> This just makes it confusing whether tests should go into libexpr-tests or tests/lang

Yup’ this should have been settled on and documented more clearly. My take on that is that we should try to write new tests in libexpr-tests (both to make the experiment meaningful and because despite the verbosity overhead writing the tests here is _so much_ faster). I’ll open a PR for that, letting you judge whether that seems a good plan.
